clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractPartitionAssignorTest.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.clients.consumer.internals;
+
+import org.apache.kafka.clients.consumer.internals.AbstractPartitionAssignor.MemberInfo;
+import org.apache.kafka.common.utils.Utils;
+import org.junit.Test;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.Random;
+
+import static org.junit.Assert.assertEquals;
+
+public class AbstractPartitionAssignorTest {
+
+    @Test
+    public void testMemberInfoSortingWithoutGroupInstanceId() {
+        MemberInfo m1 = new MemberInfo("a", Optional.empty());
+        MemberInfo m2 = new MemberInfo("b", Optional.empty());
+        MemberInfo m3 = new MemberInfo("c", Optional.empty());
+
+        List<MemberInfo> memberInfoList = Arrays.asList(m1, m2, m3);
+        assertEquals(memberInfoList, Utils.sorted(memberInfoList));
+    }
+
+    @Test
+    public void testMemberInfoSortingWithAllGroupInstanceId() {
+        MemberInfo m1 = new MemberInfo("a", Optional.of("y"));
+        MemberInfo m2 = new MemberInfo("b", Optional.of("z"));
+        MemberInfo m3 = new MemberInfo("c", Optional.of("x"));
+
+        List<MemberInfo> memberInfoList = Arrays.asList(m1, m2, m3);
+        assertEquals(Arrays.asList(m3, m1, m2), Utils.sorted(memberInfoList));
+    }
+
+    @Test
+    public void testMemberInfoSortingSomeGroupInstanceId() {
+        MemberInfo m1 = new MemberInfo("a", Optional.empty());
+        MemberInfo m2 = new MemberInfo("b", Optional.of("y"));
+        MemberInfo m3 = new MemberInfo("c", Optional.of("x"));
+
+        List<MemberInfo> memberInfoList = Arrays.asList(m1, m2, m3);
+        assertEquals(Arrays.asList(m3, m2, m1), Utils.sorted(memberInfoList));
+    }
+
+    @Test
+    public void testMergeSortManyMemberInfo() {
+        Random rand = new Random();
+        int bound = 2;
+        List<MemberInfo> memberInfoList = new ArrayList<>();
+        List<MemberInfo> staticMemberList = new ArrayList<>();
+        List<MemberInfo> dynamicMemberList = new ArrayList<>();
+        for (int i = 0; i < 100; i++) {
+            // Need to make sure all the ids are defined as 3-digits otherwise
```

Comment:
So this will only work if all group ids are the same length? Are they guaranteed to be equal length in practice?

Reply:
This is just for testing purpose. On the server side we do UUID generation which IIUC will be guaranteed to be same length.

Reply:
Here's just my readings on this:

1. For range / round-robin assignor, the first priority to still to obey its "range / round-robin" behavior, and then we may want to achieve some stickiness as possible. I.e. if achieving stickiness means we are violating the "range / round-robin" semantics then we should not do it. As a result we are only trying to be sticky with some weak efforts: i.e we only sort the consumers based on their instance ids if there are any, so that if the number of members be exactly the same and the static members do not change, we can reduce unnecessary partition migration. On the other hand, we still strictly follow the range / round-robin algorithm.

2. For sticky assignor (https://github.com/apache/kafka/pull/6963), the behavior is a bit different, but as the java docs claimed:

```
 * <p>Starting fresh it would work by distributing the partitions over consumers as evenly as possible. Even though this may sound similar to
 * how round robin assignor works, the second example below shows that it is not.
 * During a reassignment it would perform the reassignment in such a way that in the new assignment
 * <ol>
 * <li>topic partitions are still distributed as evenly as possible, and</li>
 * <li>topic partitions stay with their previously assigned consumers as much as possible.</li>
 * </ol>
 * Of course, the first goal above takes precedence over the second one.</p>
```

It still values balance over stickiness. So I think we do not need to leverage on instance.id for stickiness any more since the `owned-partitions` is a even stronger signal to do adjustments (i.e. switching partitions) in the second round.

Reply:
I partially agree here, but do we still maintain the information about `owned-partitions` when we restart a node?

Reply:
That's a good point. The `owned-partitions` are only kept in main-memory meaning that after a restart we are not sticky anymore.

I think it is okay (and straight-forward) if we want to incorporate instance.id: if no one has `owned-partitions` then we just fall back to the round-robin partitioner, and sorting based on instance.id would help; and then if any one has non-empty `owned-partitions` then that "signal" should be honored higher than instance.id, and therefore just sorting the members based on instance.ids when we do further adjustments are still sufficient.